#### Лабораторная работа 
#### Нелинейные уравнения и системы уравнений
Распад разрыва с возникновением двух ударных волн.

Пусть в начальный момент все пространство можно разделить на два однородных полупространства, граница между которыми проходит по плоскости разрыва (индекс 0 соответствует газодинамическим величинам справа от плоскости, индекс 3 - слева). Со временем в результате распада разрыва по правому и левому полупросранствам начинают распространяться две ударные волны (индекс 1 - за правой ударной волной, индекс 2 - за левой). 

Записывая три основных закона сохранения массы, количества движения и энергии на левой и правой ударных волнах и условия непрерывности давления и массовой скорости на контактном разрыве, преобразуя их, можно получить алгебраическое уравнение шестой степени вида:

\begin{equation}
    a_0Y^6 + a_1Y^5 + a_2Y^4 + a_3Y^3 + a_4Y^2 + a_5Y + a_6 = 0
\end{equation}

Выражения для коэффициентов $a_i$ приведены ниже.

$P$ - давление в среде;
$U$ - массовая скорость;
$D$ - скорость перемещения разрыва в лабораторной системе координат.

In [33]:
from numpy import *

Начальные данные

In [63]:
G0 = 1.666667
Ro0 = 7.9
P0 = 3.04 * 10**9
U0 = -2.72 * 10**4
G3 = 7.0 / 5
Ro3 = 11.37
U3 = 2.28 * 10**4
P3 = 1.17928 * 10**9

In [35]:
C3 = sqrt(G3 * (P3 / Ro3))
C0 = sqrt(G0 * (P0 / Ro0))
X = P3 / P0
alpha0 = (G0 + 1) / (G0 - 1)
alpha3 = (G3 + 1) / (G3 - 1)
e3 = 2 * C3**2 / (G3 * (G3 - 1) * (U3 - U0)**2)
e0 = 2 * C0**2 / (G0 * (G0 - 1) * (U3 - U0)**2)

In [36]:
a0 = (alpha0 * e3 - alpha3 * X * e0)**2
a0

0.06004369414360535

In [37]:
a1 = 2*((alpha0*e3-alpha3*X*e0)*(e3*(1-2*alpha0*X)-e0*X*(X-2*alpha3))-alpha0*alpha3*X*(alpha0*e3+alpha3*X*e0))
a1

-36.268390540081526

In [38]:
a2 = e3*e3*(6*alpha0*alpha0*X*X-8*alpha0*X+1)-2*e0*e3*X*(alpha0*alpha3*(X*X+4*X+1)-2*(X+1)*(alpha3+alpha0*X)+X)+e0*e0*X*X*(6*alpha3*alpha3-8*alpha3*X+X*X)+alpha0*alpha0*alpha3*alpha3*X*X-2*alpha0*X*e3*(alpha0*X-2*alpha0*alpha3*X+2*alpha3)-2*alpha3*X*X*e0*(alpha3+2*alpha0*X-2*alpha0*alpha3)
a2

125.55364653885911

In [39]:
a3 = -2*X*(2*e3*e3*(alpha0*alpha0*X*X-3*alpha0*X+1)+e0*e3*((alpha3+alpha0*X)*(X*X+4*X+1)-2*alpha0*alpha3*X*(X+1)-2*X*(X+1))+2*e0*e0*X*(X*X-3*alpha3*X+alpha3*alpha3)-alpha0*alpha3*X*(alpha0*X+alpha3)+e3*(alpha0*alpha0*alpha3*X*X-2*X*(2*alpha0*alpha3+alpha0*alpha0*X)+(2*alpha0*X+alpha3))+e0*X*(alpha0*alpha3*alpha3-2*alpha3*(alpha3+2*alpha0*X)+2*alpha3*X+alpha0*X*X))
a3

48.78769770219616

In [40]:
a4 = X*X*(e3*e3*(alpha0*alpha0*X*X-8*alpha0*X+6)-2*e0*e3*(alpha0*alpha3*X-2*(X+1)*(alpha3+alpha0*X)+X*X+4*X+1)+e0*e0*(alpha3*alpha3-8*alpha3*X+6*X*X)+(alpha3*alpha3+4*alpha0*alpha3*X+alpha0*alpha0*X*X)-2*e3*((alpha0*alpha0*X+2*alpha0*alpha3)*X-2*(2*alpha0*X+alpha3)+1)-2*e0*(alpha3*(2*alpha0*X+alpha3)-2*X*(2*alpha3+alpha0*X)+X*X))
a4

5.813227115052247

In [41]:
a5 = 2*X*X*X*(e3*e3*(alpha0*X-2)-e0*e3*(alpha0*X-2+alpha3-2*X)+e0*e0*(alpha3-2*X)+(alpha3+alpha0*X)-e3*(2*alpha0*X+alpha3-2)-e0*(2*alpha3+alpha0*X-2*X))
a5

0.09525219907497909

In [42]:
a6 = X**4*((e3-e0)*(e3-e0)+1-2*(e3+e0))
a6

-0.0061986737192730535

In [43]:
coefs = [a0, a1, a2, a3, a4, a5, a6]
coefs

[0.06004369414360535,
 -36.268390540081526,
 125.55364653885911,
 48.78769770219616,
 5.813227115052247,
 0.09525219907497909,
 -0.0061986737192730535]

In [44]:
A = max(coefs[1:])
B = max(coefs[:5])

In [45]:
A

125.55364653885911

In [46]:
B

125.55364653885911

Все корни нашего алгебраического уравнения шестой степени расположены на косплексной плоскости в кольце

\begin{equation}
    \frac{|a_n|}{|a_n| + B} \leq |z| \leq 1 + \frac{A}{|a_0|}.
\end{equation}

(Следсвие из основной теоремы алгебры комплексных чисел)

In [47]:
left = abs(a6) / (abs(a6) + B)
if a0 != 0:
    right = 1 + (A / abs(a0))
else:
    right = 1 + (A / abs(a1))

In [48]:
left

4.936828098806927e-05

In [49]:
right

2092.038007064903

Таким образом, все корни алгебраического уравнения расположены в кольце на комплексной плоскости $4.936828098806927e-05 \leq |z| \leq 2092.038007064903$

Далее, всю область делим на N частей и для дальнейшего рассмотрения оставляем только те подобласти, на границах которых $f(z)$ принимает значения разных знаков (теорема Декарта).

In [50]:
def polinom(y):
    return a0*y**6 + a1*y**5 + a2*y**4 + a3*y**3 + a4*y**2 + a5*y**1 + a6

In [51]:
len_root_area = right - left
N = 3793100
step = len_root_area / N

In [52]:
beg = []
end = []
for k in range(N):
    if sign(polinom(left + k*step)) != sign(polinom(left + (k+1)*step)):
        beg.append((left + k*step))
        end.append((left + (k+1)*step))

In [53]:
num = len(beg)

### Отрезки локализации корней

In [54]:
beg

[0.0232139550894714, 3.8464738540515295, 600.5491323693664]

In [55]:
end

[0.023765492870625765, 3.847025391832684, 600.5496839071475]

Методом дихотомии найдем значения корней уравнения

In [56]:
def dichotomy(func, x, y, eps):
    middle = (x + y) / 2
    if abs(x - y) < eps:
        return middle
    if func(middle)*func(x) < 0:
        return dichotomy(func, x, middle, eps)
    if func(middle)*func(x) > 0:
        return dichotomy(func, middle, y, eps)

In [57]:
x = []
for i in range(num):
    x.append(dichotomy(polinom, beg[i], end[i], 0.000001).round(8))

### Корни

In [58]:
x

[0.02367204, 3.84685277, 600.54916711]

### Исследуем первый корень

In [59]:
y0 = x[0]
P1 = y0 * P0
U1 = -(P1-P0)/(Ro0*C0*sqrt(((G0-1)+(G0+1)*(P1/P0))/(2*G0)))+U0
P2 = P1
U2 = U1
D0 = ((P1-P0)/Ro0+U0*U1-U0*U0)/(U1-U0)
Ro1 = (Ro0*(U0-D0))/(U1-D0)
D3 = ((P3-P2)/Ro3+U3*U3-U2*U3)/(U3-U2)
Ro2 = (Ro3*(U3-D3))/(U2-D3)
a = sqrt((P1 - P0) / (Ro1 - Ro0))
la1 = U1 / a
la0 = U0 / a

In [60]:
print("U3\t", U3)
print("D3\t", D3)

U3	 22799.999999999996
D3	 28123.41600973512


### Исследуем второй корень

In [61]:
y1 = x[1]
P1 = y1 * P0
U1 = -(P1-P0)/(Ro0*C0*sqrt(((G0-1)+(G0+1)*(P1/P0))/(2*G0)))+U0
P2 = P1
U2 = U1
D0 = ((P1-P0)/Ro0+U0*U1-U0*U0)/(U1-U0)
Ro1 = (Ro0*(U0-D0))/(U1-D0)
D3 = ((P3-P2)/Ro3+U3*U3-U2*U3)/(U3-U2)
Ro2 = (Ro3*(U3-D3))/(U2-D3)
a = sqrt((P1 - P0) / (Ro1 - Ro0))
la1 = U1 / a
la0 = U0 / a
print("U3\t", U3)
print("D3\t", D3)

U3	 22799.999999999996
D3	 10284.610379937918


### Исследуем третий корень

In [62]:
y2 = x[2]
P1 = y2 * P0
U1 = -(P1-P0)/(Ro0*C0*sqrt(((G0-1)+(G0+1)*(P1/P0))/(2*G0)))+U0
P2 = P1
U2 = U1
D0 = ((P1-P0)/Ro0+U0*U1-U0*U0)/(U1-U0)
Ro1 = (Ro0*(U0-D0))/(U1-D0)
D3 = ((P3-P2)/Ro3+U3*U3-U2*U3)/(U3-U2)
Ro2 = (Ro3*(U3-D3))/(U2-D3)
a = sqrt((P1 - P0) / (Ro1 - Ro0))
la1 = U1 / a
la0 = U0 / a
print("U3\t", U3)
print("D3\t", D3)

U3	 22799.999999999996
D3	 -321885.51358391915
